openwisp_radius/utils.py

```diff
@@ -344,41 +344,50 @@ def execute_counter_checks(
             if raise_quota_exceeded:
                 raise
 
-        if remaining is None:
+        if results is None:
             continue
-        reply_name = counter.reply_name
-        # Send remaining value in RADIUS reply, if needed.
-        # This emulates the implementation of sqlcounter in freeradius
-        # which sends the reply message only if the value is smaller
-        # than what was defined to a previous reply message
-        if reply_name not in counter_data or remaining < _get_reply_value(
-            counter_data, counter
-        ):
-            counter_data[reply_name] = remaining
+
+        # BACKWARD COMPATIBILITY: The previous implementation of counters
+        # returned a single value instead of a tuple/list when there was
+        # only one reply name defined.
+        # We need to handle this case to avoid breaking existing counters.
+        if not isinstance(results, (list, tuple)):
+            results = (results,)
```

Comment:
In the code of the counters, you check if it's a string and only then convert it to a tuple. Here you're checking if it's not a tuple or a list and convert it to a tuple. I think this inconsistent. Is it on purpose? If so, what's the reason?

Reply:
> In the code of the counters, you check if it's a string and only then convert it to a tuple

That code provides backward compatibility to the existing custom counters. 

>  Here you're checking if it's not a tuple or a list and convert it to a tuple. 

The new logic expects both `reply_names` and output of Counter.check()` to be a tuple. Thus, this code is for backward compatibility if any existing custom counter returns a single value.